src/Build.OM.UnitTests/Construction/SolutionFile_Tests.cs

```diff
@@ -16,6 +16,80 @@ namespace Microsoft.Build.UnitTests.Construction
     /// </summary>
     public class SolutionFile_Tests
     {
+        /// <summary>
+        /// Test that a project invalid project nestings is seen as invalid.
+        /// </summary>
+        [Fact]
+        public void ParseInvalidProjectNesting()
+        {
+            string solutionFileContents =
+                @"
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 15
+VisualStudioVersion = 15.0.26124.0
+MinimumVisualStudioVersion = 15.0.26124.0
+Project(""{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}"") = ""tempCode"", ""tempCode.csproj"", ""{933E47E9-37B5-4AB3-AD1C-0E48043575D7}""
+EndProject
+Global
+	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|Any CPU = Debug|Any CPU
+		Debug|x64 = Debug|x64
+		Debug|x86 = Debug|x86
+		Release|Any CPU = Release|Any CPU
+		Release|x64 = Release|x64
+		Release|x86 = Release|x86
+	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
+	GlobalSection(ProjectConfigurationPlatforms) = postSolution
+		{933E47E9-37B5-4AB3-AD1C-0E48043575D7}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+	EndGlobalSection
+        GlobalSection(NestedProjects) = preSolution
+                {933E47E9-37B5-4AB3-AD1C-0E48043575D7} = {933E47E9-37B5-4AB3-AD1C-0E48043575D7}
+        EndGlobalSection
+EndGlobal";
+            Shouldly.Should.Throw<InvalidProjectFileException>(() => ParseSolutionHelper(solutionFileContents));
+        }
+
+        /// <summary>
+        /// Test that a project invalid project nestings is seen as invalid.
+        /// </summary>
+        [Fact]
+        public void ParseInvalidProjectNesting2()
```

Comment:
I suggest adding many more graph shapes (single nodes, multiple notes, cross edges, diamond dependencies) X (with cycles, without cycles). To avoid the nuisance of specifying the whole solution each time, you can generate it instead. You can reuse / adapt the code from here: [https://github.com/cdmihai/msbuild/blob/7bcb2f961093e87277439420d351a9f3f3a8cdb8/src/Build.UnitTests/Graph/GraphLoadedFromSolution_tests.cs#L263](https://github.com/cdmihai/msbuild/blob/7bcb2f961093e87277439420d351a9f3f3a8cdb8/src/Build.UnitTests/Graph/GraphLoadedFromSolution_tests.cs#L263)

Reply:
Although I agree the testing could be more rigorous (As an alternative, I could make it into a theory and very cleanly tried a variety of different shapes), the original issue was posted almost three years ago and got no attention until I mentioned to Rainer that it looked like a quick, fun fix. I don't think it's worth taking the time to add lots of test cases to make it safer.